akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala

```diff
@@ -28,20 +28,27 @@ import akka.actor.typed.Behavior.UntypedBehavior
     ActorContextAdapter.spawnAnonymous(untyped, behavior, props)
   override def spawn[U](behavior: Behavior[U], name: String, props: Props = Props.empty) =
     ActorContextAdapter.spawn(untyped, behavior, name, props)
-  override def stop[U](child: ActorRef[U]): Boolean =
-    toUntyped(child) match {
-      case f: akka.actor.FunctionRef ⇒
-        val cell = untyped.asInstanceOf[akka.actor.ActorCell]
-        cell.removeFunctionRef(f)
-      case c ⇒
-        untyped.child(child.path.name) match {
-          case Some(`c`) ⇒
-            untyped.stop(c)
-            true
-          case _ ⇒
-            false // none of our business
-        }
+  override def stop[U](child: ActorRef[U]): Unit =
+    if (child.path.parent == self.path) { // only if a direct child
+      toUntyped(child) match {
+        case f: akka.actor.FunctionRef ⇒
+          val cell = untyped.asInstanceOf[akka.actor.ActorCell]
+          cell.removeFunctionRef(f)
+        case c ⇒
+          untyped.child(child.path.name) match {
+            case Some(`c`) ⇒
+              untyped.stop(c)
+            case _ ⇒
+            // child that was already stopped
+          }
+      }
+    } else {
+      throw new IllegalArgumentException(
```

Comment:
+1

Reply:
I kind of would want a specific error type here

Reply:
Why, this would be triggered by incorrect code, it's not like you could ever recover from it (except by changing your code, and I don't think we have a supervisor doing that ;), https://github.com/munificent/vigil )

Reply:
For understandability really. First thing I look at in failures is the exception type, it it already explains what went wrong then the message continues on to explain why etc

Reply:
I'm on the fence, it is a programming error that really is an IllegalArgument and it'll be very obvious from the message what is wrong

Reply:
`IllegalStopOfNonChildException extends IAE`